README.md
# oai_kpa_mpp

CHANGELOG.md
# Changelog

<!--next-version-placeholder-->

## v0.2.0 (2021-11-25)
### Feature
* Added mpp impact feature ([`56d4b18`](https://github.com/CrinitusFeles/oai_kpa_mpp/commit/56d4b186bd8fcec1d9646252c2561fcd04d220e0))

## v0.1.2 (2021-08-31)
### Fix
* Added oai_modbus dependency to .toml ([`187339a`](https://github.com/CrinitusFeles/oai_kpa_mpp/commit/187339a3ba13e5911fd8b5d6e08e697bf71bb90e))

## v0.1.1 (2021-08-31)
### Fix
* Added semantic releases ([`f0c0fc9`](https://github.com/CrinitusFeles/oai_kpa_mpp/commit/f0c0fc9d9e964ecb0ca1d34f00706ba8765dd510))


oai_kpa_stm/oai_kpa_mpp_test.py
import oai_modbus
import threading
import time

if __name__ == '__main__':
   #mpp = oai_modbus.OAI_Modbus(serial_num=['207F369F424D'], debug=True)
    mpp = oai_modbus.OAI_Modbus(serial_num=['206D3699424D'], debug=True)

    mpp.connect()
    """
    GPIO MAP:

    DEP GPIOs :     GPIO 1 = GPIO 1     //      DEP voltage turn-on/turn-off
                    GPIO 3 = GPIO 3     //      -30 voltage turn-on
                    GPIO 7 = GPIO 7     //      +30 voltage turn-on
       """


    def GPIO_initialization():
        """
        Initialization of certain GPIO groups
        :param target: all_slots, usb hubs, external available GPIO pins, all GPIO pins
        :return:    nothing
        """
        mpp.write_regs(1060, [0x1a30, 0x0000, 0x0000, 0x0000])  # GPIOs 1,3,7 initialization


    def DEP_control():
        """
        Set value for certain GPIO groups
        :param data:
        :param value:
        :param target: all_slots, usb hubs, external available GPIO pins, all GPIO pins
        :return: nothing
        """
        mpp.write_regs(1060, [0x1a30, 0x0000, 0x0000, 0x0000])  # GPIOs 1,7 initialization
        while True:
            mpp.write_regs(1064, [0x1a00, 0x0000, 0x0000, 0x0000])  # GPIOs 1,3 set //  current in the coil
            mpp.write_regs(1064, [0x1300, 0x0000, 0x0000, 0x0000])  # GPIOs 1 set //  -30 voltage turn-on
            time.sleep(3)
            mpp.write_regs(1064, [0x1330, 0x0000, 0x0000, 0x0000])  # GPIOs 1,7 set //  current in the coil
            mpp.write_regs(1064, [0x1300, 0x0000, 0x0000, 0x0000])  # GPIOs 1 set //  +30 voltage turn-on
            time.sleep(3)
            mpp.write_regs(1064, [0x1000, 0x0000, 0x0000, 0x0000])  # GPIOs 1,7 set //  DEP voltage turn-off
            time.sleep(3)



    time.sleep(0.1)
    GPIO_initialization()
    DEP_control()

    mpp.disconnect()


oai_kpa_mpp/oai_kpa_mpp.py
# -*- coding: utf-8 -*-

# модули GUI
from PyQt5 import QtWidgets, QtCore, QtGui
from PyQt5 import QtTest
import sys
# стандартные модули для работы
import time
import os
import re
import json
# модули ОАИ_КПА
from PyQt5.QtWidgets import QApplication

try:
    from . import oai_kpa_mpp_widget_qt
    from . import oia_kpa_mpp_data
    from . import oai_kpa_mpp_impact
except ImportError:
    print('starting as module')
    import oai_kpa_mpp_widget_qt
    import oia_kpa_mpp_data
    import oai_kpa_mpp_impact


class ClientGUIWindow(QtWidgets.QWidget, oai_kpa_mpp_widget_qt.Ui_Form):
    def __init__(self, *args, **kwargs):
        # # Стандартная часть окна # #
        # обязательная часть для запуска Qt-виджета
        super().__init__()
        self.setupUi(self)
        # создание и обработка словаря настройки (здесь же обрабатывается параметры **kwargs)
        self.uniq_name = kwargs.get("uniq_name", 'oai_kpa_mpp_un')
        # настройки по умолчанию
        # настройки не для изменения (одинаковые для каждого типа плат)
        self.core_cfg = {'serial_num': '207F369F424D',
                         'widget': True}
        # настройки для вашего модуля (разные для каждого типа плат)
        self.channels_default_parameters = {num: "АЦП %d К %d" % (num//16, num%16)
                                            for num in range(32)}
        self.user_cfg = {"channels": self.channels_default_parameters}
        self.default_cfg = {'core': self.core_cfg,
                            'user': self.user_cfg
                            }
        #
        self.loaded_cfg = self.load_cfg()
        self.cfg = self.cfg_process(self.loaded_cfg, kwargs)
        # скрываем ненужные элементы
        if self.cfg["core"]["widget"] is str(True):
            self.connectionPButton.hide()

        # описываем элементы стандартного окна
        self.connectionPButton.clicked.connect(self.reconnect)

        # переменные для создание лога
        self.log_file_title = self.generate_log_title()
        self.log_file_data = ["0" for i in range(len(self.log_file_title))]
        self.log_file = None
        self.recreate_log_files()

        # отслеживание состояния окна
        self.state = 0

        # настройки отображения и сохранения в лог
        self.gui_update_time_ms = 1000
        self.log_update_time_ms = 1000

        # Таймер для создания псевдопотока для обновления GUI
        self.data_update_timer = QtCore.QTimer()
        self.data_update_timer.timeout.connect(self.update_gui)
        self.data_update_timer.start(self.gui_update_time_ms)

        # Таймер для создания псевдопотока для установления поля ДЭП
        self.DEP_data_update_timer = QtCore.QTimer()
        self.DEP_data_update_timer.timeout.connect(self.voltage_cycling)

        # Таймер для создания псевдопотока для сохранения данных в лог
        self.log_update_timer = QtCore.QTimer()
        self.log_update_timer.timeout.connect(self.log_write)
        self.log_update_timer.start(self.log_update_time_ms)
        # # Изменяемая часть окна # #

        self.moduleSerialNumberLEdit.setText(self.cfg["core"]["serial_num"])

        # Часть под правку: здесь вы инициализируете необходимые компоненты
        self.module = oia_kpa_mpp_data.OaiKpaMPP(serial_num=self.cfg["core"]["serial_num"])
        self.impact_mpp_ch1 = oai_kpa_mpp_impact.ImpactModuleMPP(channel=1)
        self.impact_mpp_ch2 = oai_kpa_mpp_impact.ImpactModuleMPP(channel=2)
        self.impact_mpp_ch1.set_client(self.module.client)
        self.impact_mpp_ch2.set_client(self.module.client)
        self.en_ch1_pushbutton.pressed.connect(self.start_ch1_impact)
        self.en_ch2_pushbutton.pressed.connect(self.start_ch2_impact)

        # the table for stm_channels data visualisation
        self.stm_color_map = {0: "darkturquoise", 1: "darkseagreen", 2: "lightcoral"}
        self.stm_table_column, self.stm_table_row = 4, 8
        self.table_values = [[{"voltage": 0.0, "color": "gray"}for i in range(self.stm_table_row)]
                             for j in range(self.stm_table_column)]

        # Кнопки для управления модулем
        self.turn_on_minus_30V_.clicked.connect(self.DEP_button_pressed)
        self.turn_on_0V_.clicked.connect(self.DEP_button_pressed)
        self.turn_on_30V_.clicked.connect(self.DEP_button_pressed)


        # циклическая подачи напряжения ДЭП
        self.dep_state = 0
        self.cycle_DEP_start_.clicked.connect(self.voltage_cycling_start)
        self.cycle_DEP_stop_.clicked.connect(self.voltage_cycling_stop)
        self.cycle_period = self.cycle_window_.value()

        self.cycle_reading_flag = False

    def DEP_button_pressed(self):
        if self.sender().text() == '30 В':
            self.dep_state = 1
            self.module.DEP_control(30)
        elif self.sender().text() == "-30 В":
            self.dep_state = 2
            self.module.DEP_control(-30)
        elif self.sender().text() == "0 В":
            self.dep_state = 0
            self.module.DEP_control(0)

    def voltage_cycling_start(self):
        self.cycle_period = self.cycle_window_.value()
        #print(self.cycle_period)
        self.DEP_data_update_timer.start(1000*self.cycle_period)

    def voltage_cycling_stop(self):
        self.DEP_data_update_timer.stop()

    def voltage_cycling(self):

        self.dep_state += 1
        self.dep_state %= 3

        #print(self.dep_state)
        if self.dep_state == 0:
            self.module.DEP_control(0)
            pass
        if self.dep_state == 1:
            self.module.DEP_control(30)
            pass
        if self.dep_state == 2:
            self.module.DEP_control(-30)
            pass
        #QtTest.QTest.qWait(1000)
        #print("world")
        #QtTest.QTest.qWait(1000)

    def connection_state_check(self):
        """
        The useful method to generate correct status string with color
        """
        if self.module.state == -2:
            self.set_status_string(string="Ошибка подключения", color="lightcoral")
        elif self.module.state == -1:
            self.set_status_string(string="Ошибка подключения", color="orangered")
        elif self.module.state == 0:
            self.set_status_string(string="Необходимо подключение", color="white")
        elif self.module.state == 1:
            self.set_status_string(string="Подключение успешно", color="darkseagreen")
        else:
            self.set_status_string(string="Подключение успешно", color="white")
        pass

    def set_status_string(self, string="Нет информации", color="white"):
        """
        setting string and color to gui output status-line
        :param string: string to GUI-output
        :param color: background color
        """
        self.statusLineEdit.setText(str(string))
        self.statusLineEdit.setStyleSheet('QLineEdit {background-color: %s;}' % color)

    def connect(self):
        """
        connection to kpa-module
        :return: nothing
        """
        serial_number = self.moduleSerialNumberLEdit.text()
        if re.findall(r"[0-9a-fA-F]{8,12}", serial_number):
            self.cfg["core"]["serial_num"] = serial_number
        else:
            serial_number = self.serial_number
            self.moduleSerialNumberLEdit.setText(self.cfg["core"]["serial_num"])
        self.module.connect(serial_num=serial_number)
        self.connection_state_check()
        #
        self.save_cfg()
        # self.impact_mpp_ch1.set_client(self.module.client)
        # self.impact_mpp_ch2.set_client(self.module.client)
        pass

    def disconnect(self):
        """
        disconnect from kpa_module, if connection is established; in other cases do nothing
        :return: nothing
        """
        self.module.disconnect()
        self.connection_state_check()
        pass

    def reconnect(self):
        """
        reconnect module
        :return: nothing
        """
        self.disconnect()
        self.connect()
        self.connection_state_check()
        pass

    def update_gui(self):
        """
        the function, that update gui data output parts (table, etc). Run each <gui_update_time_ms>.
        :return: nothing
        """
        if self.cycle_reading_flag:
            self.fill_table_data_from_stm_data()
        pass

    def cycle_reading(self):
        """
        visualisation the status of cycle reading
        :return: nothing
        """
        if self.cycle_reading_flag:
            self.cycle_reading_flag = False
            color = "lightgray"
        else:
            self.cycle_reading_flag = True
            color = "darkseagreen"
        self.cycleReadPushButton.setStyleSheet('QPushButton {background-color: %s;}' % color)
        pass

    def fill_table_data_from_stm_data(self):
        """
        filling the stm-table
        :return:
        """
        try:
            if self.module.state == 1:
                for column in range(self.stm_table_column):
                    for row in range(self.stm_table_row):
                        adc_num, ch_num = column // 2, row + self.stm_table_row*(column % 2)
                        value, state = self.module.get_channel_values(adc_num, ch_num)
                        self.__fill_single_socket(self.stmTableWidget, row, 2*column+1, value,
                                                  color=self.stm_color_map.get(state, "white"))
                        name = self.cfg["user"]["channels"][str(adc_num*16 + ch_num)]
                        self.__fill_single_socket(self.stmTableWidget, row, 2*column, name,
                                                  color="white")
            else:
                pass
        except Exception as error:
            # print("fill_table_data ", error)
            pass

    @staticmethod
    def __fill_single_socket(table, row, column, value, color=None):
        """
        fill the socket of table by value and color
        :param table: the table for filling
        :param row: the table row of item for filling
        :param column: the table column of item for filling
        :param value: value to put in table socket
        :param color: color of table item
        :return: nothing
        """
        if type(value) == str:
            table_item = QtWidgets.QTableWidgetItem(value)
        elif type(value) == float:
            table_item = QtWidgets.QTableWidgetItem("%.3f V" % value)
        else:
            table_item = QtWidgets.QTableWidgetItem("%s" % value)
        table_item.setTextAlignment(QtCore.Qt.AlignHCenter | QtCore.Qt.AlignVCenter)
        if color:
            table_item.setBackground(QtGui.QColor(color))
        table.setItem(row, column, table_item)
        pass

    # работа с файлом конфигурации
    @staticmethod
    def cfg_process(default_cfg, new_cfg):
        """
        Process default and new cfg-s and forms actual cfg
        :param default_cfg: default parameters set
        :param new_cfg: cfg to update
        :return: actual_cfg
        """
        cfg = default_cfg
        for key, value in new_cfg.items():
            for c_key, c_value in default_cfg["core"].items():
                if c_key == key:
                    cfg["core"][key] = value
            for c_key, c_value in default_cfg["user"].items():
                if c_key == key:
                    cfg["user"][key] = value
        return cfg

    def save_cfg(self):
        try:
            os.mkdir("cfg")
        except OSError as error:
            pass
        #
        with open("cfg\\" + self.uniq_name + ".json", 'w', encoding="utf-8") as cfg_file:
            json.dump(self.cfg, cfg_file, sort_keys=True, indent=4, ensure_ascii=False)

    def save_default_cfg(self):
        try:
            os.mkdir("cfg")
        except OSError as error:
            pass
        #
        with open("cfg\\" + self.uniq_name + ".json", 'w', encoding="utf-8") as cfg_file:
            json.dump(self.default_cfg, cfg_file, sort_keys=True, indent=4, ensure_ascii=False)

    def load_cfg(self):
        try:
            with open("cfg\\" + self.uniq_name + ".json", 'r', encoding="utf-8") as cfg_file:
                loaded_cfg = json.load(cfg_file)
        except FileNotFoundError:
            loaded_cfg = self.default_cfg
        return loaded_cfg

    # работа с log-файлом
    @staticmethod
    def create_log_file(file=None, dir_name="log", sub_dir="log", sub_sub_dir=True, prefix="", extension=".csv"):
        """
        log-file creation
        :param file: if log-file already created, it will be closed
        :param dir_name: the folder, where logs will be stored
        :param sub_dir: the postfix for time_date (%Y_%m_%d_<sub_dir>) sub_dir_name
        :param sub_sub_dir: if True in sub_dir the log-files will be placed in additional folder (%Y_%m_%d_%H-%M-%S_<sub_dir>)
        :param prefix: the file-name prefix ("%Y_%m_%d %H-%M-%S <prefix> <extension>)
        :param extension: the file-name extension ("%Y_%m_%d %H-%M-%S <prefix> <extension>)
        :return: lof_file handle
        """
        sub_dir_name = dir_name + "\\" + time.strftime("%Y_%m_%d", time.localtime()) + " " + sub_dir
        if sub_sub_dir:
            sub_sub_dir_name = sub_dir_name + "\\" + time.strftime("%Y_%m_%d %H-%M-%S ",
                                                                   time.localtime()) + sub_dir
        else:
            sub_sub_dir_name = sub_dir_name
        try:
            os.makedirs(sub_sub_dir_name)
        except (OSError, AttributeError) as error:
            print(error)
            pass
        try:
            if file:
                file.close()
        except (OSError, NameError, AttributeError) as error:
            print(error)
            pass
        file_name = sub_sub_dir_name + "\\" + time.strftime("%Y_%m_%d %H-%M-%S ",
                                                            time.localtime()) + prefix + " " + extension
        file = open(file_name, 'a', encoding="utf-8")
        return file

    @staticmethod
    def close_log_file(file=None):
        """
        closing lo-file, if it possible; in other cases does nothing
        :param file: file to close
        """
        if file:
            try:
                file.close()
            except (OSError, NameError, AttributeError) as error:
                print(error)
            finally:
                file = None
        pass

    def recreate_log_files(self):
        """
        log-file recreation
        """
        # перезапуск лог файла
        self.log_file = self.create_log_file(file=self.log_file, dir_name="log", sub_dir=self.uniq_name,
                                             sub_sub_dir=False, prefix=self.uniq_name, extension=".csv")
        self.log_file.write(self.log_file_title)
        pass

    def generate_log_title(self):
        """
        log-file title list generation
        :return: title list for log-file
        """
        # обязательная часть - время в формате с.мс
        log_title = ["Time, s"]
        # список данных, генерируемых модулем
        data_title_list = [self.cfg["user"]["channels"].get(str(i), str(i)) for i in range(32)]
        #
        log_title.extend(data_title_list)
        return ";".join(log_title) + "\n"

    def generate_log_data(self):
        """
        log-file data list generation
        :return: data list for log-file
        """
        # обязательная часть - время в формате с.мс
        log_title = ["%.3f" % time.perf_counter()]
        # список данных, генерируемых модулем
        data_title_list = ["%.3f" % value for value in self.module.get_channels_values()[0]]
        #
        log_title.extend(data_title_list)
        return ";".join(log_title) + '\n'

    def log_write(self):
        """
        function, witch rerun every log_update_time_ms, to write data to log
        :return: nothing
        """
        log_data = self.generate_log_data()
        if log_data:
            self.log_file.write(log_data)
        pass

    # обработка события закрытия окна
    def closeEvent(self, event):
        self.save_cfg()
        pass

#  ------------------------  MPP impact module  ----------------------------
    def start_ch1_impact(self):
        try:
            self.impact_mpp_ch1.init_impact_parameters(U0=self.ch1_u0_spinbox.value(), U1=self.ch1_u1_spinbox.value(),
                                                       N=self.ch1_N_spinbox.value(), M=self.ch1_M_spinbox.value(),
                                                       T=self.ch1_T_spinbox.value(), t=self.ch1_t_spinbox.value())
            self.impact_mpp_ch1.write_impact_parameters()
            print(self.impact_mpp_ch1.get_impact_parameters())
            self.impact_mpp_ch1.start_impact()
        except Exception as ex:
            print(ex)
    
    def start_ch2_impact(self):
        try:
            self.impact_mpp_ch2.init_impact_parameters(U0=self.ch2_u0_spinbox.value(), U1=self.ch2_u1_spinbox.value(),
                                                       N=self.ch2_N_spinbox.value(), M=self.ch2_M_spinbox.value(),
                                                       T=self.ch2_T_spinbox.value(), t=self.ch2_t_spinbox.value())
            self.impact_mpp_ch2.write_impact_parameters()
            print(self.impact_mpp_ch2.get_impact_parameters())
            self.impact_mpp_ch2.start_impact()
        except Exception as ex:
            print(ex)
        

# ==========================================================================
if __name__ == "__main__":
    app = QApplication(sys.argv)
    w = ClientGUIWindow(serial_num='207F369F424D')
    w.show()
    app.exec_()

oai_kpa_mpp/oia_kpa_mpp_data.py
# -*- coding: utf-8 -*-

import time
import threading
try:
    import OAI_Modbus as oai_modbus
except ImportError:
    import oai_modbus
import json


class CfgParameter:
    def __init__(self, **kwargs):
        self.serial_number = kwargs.get('serial_num', '206D3699424D')


class OaiKpaMPP:
    def __init__(self, *args, **kwargs):
        # объект с параметрами
        self.cfg = CfgParameter(serial_num="207F369F424D")
        
        # разбор именованных параметров
        self.serial_number = kwargs.get('serial_num', '207F369F424D')
        self.debug = kwargs.get('debug', False)
        # создание объекта для общения по МодБас
        self.client = oai_modbus.OAI_Modbus(serial_num=[self.serial_number])
        self.client.debug_print_flag = True
        # параметры связи
        self.state = 0
        # описание связи через запись/чтение регистров
        """
            mode: master - 0, slave - 1
            direction: 2Line - 0, 2Line_rx_only - 1, 1Line - 2
            data_size: 8 bit - 0, 16 bit - 1
            polarity: low - 0, high - 1
            phase: one_edge - 0, two_edge - 1
            slave: not_used_et
            baud: 0 - 21 MB / s, 1 - 10, 5 MB / s, 2 - 5, 25 MB / s, 3 - 2, 625 MB / s, 4 - 1, 3125 MB / s, 5 - 656, 25 kB / s, 6 - 328, 125 kB / s, 7 - 164, 062 kB / s.
            firs_bit: 0 - MSB_first, 1 - LSB_first 
            ti_mode: 0 - disable, 1 – enable
        """
        self.register_addr = {
            "scaler": 1246,
            "mode": 1247,
            "direction": 1248,
            "polarity": 1249,
            "phase": 1250,
            "slave": 1251,
            "baud": 1252,
            "first_bit": 1253,
            "ti_mode": 1254
        }
        # параметры работы модуля
        self.iteration = 0
        self.adc_param = {"adc_num": 2, "channel_num": 16}
        self.channel_row_adc_data = [[0 for i in range(self.adc_param["channel_num"])]
                                     for j in range(self.adc_param["adc_num"])]
        self.channel_adc_voltage = [[0 for i in range(self.adc_param["channel_num"])]
                                    for j in range(self.adc_param["adc_num"])]
        self.adc_cal_a = [[1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0],
                          [1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0]]
        self.adc_cal_b = [[0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0],
                          [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0]]
        self.stm_max_min_bound = {"min": [[1000.0, 1000.0, 1000.0, 1000.0, 1000.0, 1000.0, 1000.0, 1000.0,
                                          1000.0, 1000.0, 1000.0, 1000.0, 1000.0, 1000.0, 1000.0, 1000.0],
                                          [1000.0, 1000.0, 1000.0, 1000.0, 1000.0, 1000.0, 1000.0, 1000.0,
                                          1000.0, 1000.0, 1000.0, 1000.0, 1000.0, 1000.0, 1000.0, 1000.0]],
                                  "max": [[1500.0, 1500.0, 1500.0, 1500.0, 1500.0, 1500.0, 1500.0, 1500.0,
                                          1500.0, 1500.0, 1500.0, 1500.0, 1500.0, 1500.0, 1500.0, 1500.0],
                                          [1500.0, 1500.0, 1500.0, 1500.0, 1500.0, 1500.0, 1500.0, 1500.0,
                                           1500.0, 1500.0, 1500.0, 1500.0, 1500.0, 1500.0, 1500.0, 1500.0]]
                                  }
        self.channel_state = [[0 for i in range(self.adc_param["channel_num"])]
                              for j in range(self.adc_param["adc_num"])]  # (0 - open, 1 - short, 2 - undefined)
        #
        #self.read_write_thread = threading.Thread(target=self.stm_update, args=(), daemon=True)
        #self.read_write_thread.start()
        self.adc_data_lock = threading.Lock()

    def connect(self, serial_num=None):
        """
        connection to the HW-module
        connection parameter can be updated
        :param serial_num: serial_number
        :return: nothing
        """
        if serial_num:
            self.serial_number = serial_num
            self.client.serial_numbers.append(self.serial_number)
        pass
        if self.client.connect() == 1:
            self.state = 1
        else:
            self.state = -1
        return self.state

    def disconnect(self):
        try:
            if self.client.disconnect() == 0:
                self.state = 0
            else:
                self.state = -1
        except AttributeError:
            self.state = -1
            pass
        return self.state

    def stm_update(self):
        while 1:
            if self.client.connection_status:
                #
                self.client.start_continuously_queue_reading(ai=[[2074, 2078]], ao=[], write=[])
                #
                self.client.write_regs(offset=1246, data_list=[1, 0, 0, 1,
                                                               0, 1, 0, 6,
                                                               0, 0])
                self.client.write_regs(offset=1318, data_list=[1, 0, 0, 1, 2])
                #
                self.iteration += 1
                ch_num = self.iteration % self.adc_param["channel_num"]
                adc_num = (self.iteration//self.adc_param["channel_num"]) % self.adc_param["adc_num"]
                # set cs
                # set control register
                control_register = ((0x02 << 10) | ((ch_num & 0x0F) << 6) | (0x03 << 4) | (0x01 << 2) | (0x01 << 0)) << 4
                self.client.write_regs(offset=1276, data_list=[control_register, 0x00])

                if adc_num == 0:
                    self.client.write_regs(offset=1266, data_list=[1, 0, 1, 1, 0, 1, 1])
                else:
                    self.client.write_regs(offset=1266, data_list=[1, 0, 1, 1, 0, 1, 2])
                # release cs
                # stm_mod.client.write_regs(offset=1060, data_list=[0x1C00, 0x0000, 0x0000, 0x0000])
                # stm_mod.client.write_regs(offset=1064, data_list=[0x1C00, 0x0000, 0x0000, 0x0000])
                #
                time.sleep(0.1)

                channel_num = (self.client.ai_register_map[2074] >> 12) & 0x0F
                channel_data = (self.client.ai_register_map[2074] & 0xFFF) << 0

                print("%04X" % self.client.ai_register_map[2074])

                if channel_num != ch_num:
                    # raise ValueError("Error with adc channel")
                    pass
                self.channel_row_adc_data[adc_num][channel_num] = channel_data & 0xFFF
                self.__refresh_channel_values(adc_num, ch_num)
        pass

    def DEP_control(self, voltage):
        """
        :param voltage: voltage
        :return: nothing
        """
        self.client.write_regs(1060, [0x1a30, 0x0000, 0x0000, 0x0000])  # GPIOs 1,7 initialization

        if voltage == 30:
            self.client.write_regs(1064, [0x1330, 0x0000, 0x0000, 0x0000])  # GPIOs 1,7 set //  current in the coil
            self.client.write_regs(1064, [0x1300, 0x0000, 0x0000, 0x0000])  # GPIOs 1 set //  +30 voltage turn-on

        elif voltage == -30:
            self.client.write_regs(1064, [0x1a00, 0x0000, 0x0000, 0x0000])  # GPIOs 1,3 set //  current in the coil
            self.client.write_regs(1064, [0x1300, 0x0000, 0x0000, 0x0000])  # GPIOs 1 set //  -30 voltage turn-on

        elif voltage == 0:
            self.client.write_regs(1064, [0x1000, 0x0000, 0x0000, 0x0000])  # GPIOs 1,7 set //  DEP voltage turn-off


    def get_channel_values(self, adc, ch_num):
        """
        return state of single adc_channel
        :param adc: adc number (0 or 1)
        :param ch_num: adc channel number (from 0 to 15)
        :return: value pair of voltage and state (0 - open, 1 - short, 2 - undefined)
        """
        with self.adc_data_lock:
            try:
                return self.channel_adc_voltage[adc][ch_num], self.channel_state[adc][ch_num]
            except IndexError:
                raise IndexError("Incorrect adc (0, 1) or channel number(0-15")

    def get_channels_values(self):
        """
        return state of all adc_channels
        :return: list of value pairs of voltage and state (0 - open, 1 - short, 2 - undefined)
        """
        with self.adc_data_lock:
            try:
                val_list = [self.channel_adc_voltage[adc][ch_num] for ch_num in range(self.adc_param["channel_num"])
                            for adc in range(self.adc_param["adc_num"])]
                state_list = [self.channel_state[adc][ch_num] for ch_num in range(self.adc_param["channel_num"])
                              for adc in range(self.adc_param["adc_num"])]
                return val_list, state_list
            except IndexError:
                raise IndexError("Incorrect adc (0, 1) or channel number(0-15")

    def __refresh_channel_values(self, adc, ch_num):
        with self.adc_data_lock:
            a = self.adc_cal_a[adc][ch_num]
            b = self.adc_cal_b[adc][ch_num]
            self.channel_adc_voltage[adc][ch_num] = a * self.channel_row_adc_data[adc][ch_num] + b
            self.channel_state[adc][ch_num] = self.__get_state_from_bound(self.channel_adc_voltage[adc][ch_num],
                                                                          self.stm_max_min_bound["min"][adc][ch_num],
                                                                          self.stm_max_min_bound["max"][adc][ch_num])

    def data_process(self):
        """
        calculate all data from adc_value
        :return: nothing
        """
        for adc in range(self.adc_param["adc_num"]):
            for ch_num in range(self.adc_param["channel_num"]):
                a = self.adc_cal_a[adc][ch_num]
                b = self.adc_cal_b[adc][ch_num]
                self.channel_adc_voltage[adc][ch_num] = a*self.channel_row_adc_data[adc][ch_num] + b
                self.channel_state[adc][ch_num] = self.__get_state_from_bound(self.channel_adc_voltage[adc][ch_num],
                                                                 self.stm_max_min_bound["min"][adc][ch_num],
                                                                 self.stm_max_min_bound["max"][adc][ch_num])

    @staticmethod
    def __get_state_from_bound(val, bound_1, bound_2):
        """
        value checks on correspondence to bounds intervals
        :param val:  value on test
        :param bound_1: minimal bound
        :param bound_2: maximum bound
        :return: state: 1 - val <= bound_1, 0 - val >= bound_2, 2 - other
        """
        if val >= bound_2:
            return 0
        elif val <= bound_1:
            return 1
        else:
            return 2

    def __repr__(self):
        return_str = "oai_kpa_stm module: ser_num %s\n" % self.serial_number
        for adc in range(self.adc_param["adc_num"]):
            return_str += "\t <adc %02d> " % adc
            for ch_num in range(self.adc_param["channel_num"]):
                return_str += "\t %02d: %04.3f" % (ch_num, self.channel_adc_voltage[adc][ch_num])
            return_str += "\n"
        return return_str


if __name__ == '__main__':
    itteration_num = 0
    stm_mod = OaiKpaSTM(serial_num="20713699424D", debug=False)
    print("Connect")
    stm_mod.connect()
    #
    while 1:
        print(stm_mod)
        time.sleep(1)

